monai/data/wsi_datasets.py

```diff
@@ -135,3 +136,87 @@ def _transform(self, index: int):
         # Create put all patch information together and apply transforms
         patch = {"image": image, "label": label, "metadata": metadata}
         return apply_transform(self.transform, patch) if self.transform else patch
+
+
+class SlidingPatchWSIDataset(PatchWSIDataset):
+    """
+    This dataset extracts patches from whole slide images (without loading the whole image)
+    It also reads labels for each patch and provides each patch with its associated class labels.
+
+    Args:
+        data: the list of input samples including image, location, and label (see the note below for more details).
+        size: the size of patch to be extracted from the whole slide image.
+        level: the level at which the patches to be extracted (default to 0).
+        transform: transforms to be executed on input data.
+        reader: the module to be used for loading whole slide imaging. Defaults to cuCIM. If `reader` is
+
+            - a string, it defines the backend of `monai.data.WSIReader`.
+            - a class (inherited from `BaseWSIReader`), it is initialized and set as wsi_reader,
+            - an instance of a a class inherited from `BaseWSIReader`, it is set as the wsi_reader.
+
+        kwargs: additional arguments to pass to `WSIReader` or provided whole slide reader class
+
+    Note:
+        The input data has the following form as an example:
+
+        .. code-block:: python
+
+            [
+                {"image": "path/to/image1.tiff"},
+                {"image": "path/to/image2.tiff", "size": [20, 20], "level": 2}
+            ]
+
+    """
+
+    def __init__(
+        self,
+        data: Sequence,
+        size: Optional[Union[int, Tuple[int, int]]] = None,
+        level: Optional[int] = None,
+        overlap: float = 0,
+        transform: Optional[Callable] = None,
+        reader="cuCIM",
+        **kwargs,
+    ):
+        super().__init__(data=data, size=size, level=level, transform=transform, reader=reader, **kwargs)
+        self.overlap = overlap
+
+        # Create single sample for each patch (in a sliding window manner)
+        self.data = []
+        for sample in data:
+            sliding_samples = self._make_patches(sample)
+            self.data.extend(sliding_samples)
```

Comment:
could you please consider the existing implementations for patch analysis? https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/grid_dataset.py

Reply:
Thanks for your suggestion, @wyli! Taking a quick look at `iter_patch`, I think it is not generalizable to WSI patches. For its functionalities and some sub-routines, it requires the entire image as an input to work, right?
https://github.com/Project-MONAI/MONAI/blob/209db9e08129855df878634639d4c2700d9acd83/monai/data/utils.py#L192-L199

For WSI we don't have the image array in the memory although we have some information like the dimension and size. Do you think without the image array, we can leverage the existing patch analysis implementations?

Reply:
sure, there are also https://github.com/Project-MONAI/MONAI/blob/209db9e08129855df878634639d4c2700d9acd83/monai/data/utils.py#L122
and https://github.com/Project-MONAI/MONAI/blob/209db9e08129855df878634639d4c2700d9acd83/monai/data/utils.py#L152
and the iterator avoids saving all the coordinates in memory, looks quite efficient.

the patch-based analysis is important for monai. if the existing utilities don't look good, could you please help improve and adapt?

Reply:
@wyli, I checked these functions and non of them are re-usable for whole slide patches.
There are two fundamental differences: 
|  | Array Patch | WSI Patch
| -- | -- | --
| patch definition | array slices | starting location and size 
| padding | any np.pad mode | handled by 3rd party WSI reader libraries and generally limited to constant values |
| resolution | there is no notion of resolution | patches can be defined on any resolution level but always need to be calculated in level=0

Also here in this dataset we don't want to have iterators, since we want the size of the dataset to be defined based on number of patches and usually there is no problem with having even millions of patch locations in memory.

Reply:
I'll make a similarly named function `iter_wsi_patch_location` in `monai.data.utils` and will use it in this dataset but it will be different than the other patch function.

Reply:
thanks for the summary, in the latest version `downsample` is not varying within `iter_wsi_patch_location`, so can we precompute `patch_size_` before line 203 here: https://github.com/Project-MONAI/MONAI/blob/1700fdab7972221efcc87baa963db68d24ff4b3a/monai/data/wsi_datasets.py#L203-L204
so that we avoid `iter_wsi_patch_location` (because duplicating the code adds maintenance efforts). Looks like `padded: bool` could be part of `iter_patch` as well?

Reply:
That's a good idea! However, since `iter_patch_slices` return `slice` while here we just need the starting position, I will modify it to have one function `iter_patch_position`.

Reply:
I made some modification to remove any redundancy. Now we have `iter_patch_position` that returns the upper left corner of each patch (this is what we need for wsi patches), and `iter_patch_slices` that calls `iter_patch_position` and creates slices using the positions.

Reply:
@wyli let me know if you have any other comments. Thanks

Reply:
thanks, the new impl. are more generic than the current `dev`'s, that looks good to me.